saleor/plugins/avatax/plugin.py

```diff
@@ -478,35 +479,62 @@ def calculate_order_line_total(
 
+        tax_included = (
+            lambda: Site.objects.get_current().settings.include_taxes_in_prices
+        )
+
         taxes_data = self._get_order_tax_data(order, previous_value)
         return self._calculate_order_line_total_price(
-            taxes_data, variant.sku, previous_value
+            taxes_data, variant.sku, tax_included, previous_value
         )
 
     @staticmethod
     def _calculate_order_line_total_price(
         taxes_data: Dict[str, Any],
         item_code: str,
+        tax_included: Callable[[], bool],
         base_value: TaxedMoney,
-    ):
+    ) -> OrderTaxedPricesData:
         if not taxes_data or "error" in taxes_data:
             return base_value
 
-        tax_included = (
-            lambda: Site.objects.get_current().settings.include_taxes_in_prices
-        )
         currency = taxes_data.get("currencyCode")
+        undiscounted_line_price = None
+        line_price_with_discounts = None
+
         for line in taxes_data.get("lines", []):
-            if line.get("itemCode") == item_code:
-                tax = Decimal(line.get("tax", 0.0))
-                net = Decimal(line["lineAmount"])
-                if currency == "JPY" and tax_included():
-                    line_gross = base_value.gross
-                    line_net = Money(amount=line_gross.amount - tax, currency=currency)
-                else:
-                    line_gross = Money(amount=net + tax, currency=currency)
-                    line_net = Money(amount=net, currency=currency)
-                return TaxedMoney(net=line_net, gross=line_gross)
+            if line.get("itemCode") != item_code:
+                continue
+            is_discount_record = line.get("ref1")
+
+            tax = Decimal(line.get("tax", 0.0))
+            net = Decimal(line["lineAmount"])
+
+            if currency == "JPY" and tax_included():
+                line_gross = base_value.undiscounted_price.gross
+                if is_discount_record:
+                    line_gross = base_value.price_with_discounts.gross
+                line_net = Money(amount=line_gross.amount - tax, currency=currency)
+            else:
+                line_gross = Money(amount=net + tax, currency=currency)
+                line_net = Money(amount=net, currency=currency)
+
+            total = TaxedMoney(net=line_net, gross=line_gross)
+            if is_discount_record:
+                line_price_with_discounts = total
+            else:
+                undiscounted_line_price = total
+
+        if undiscounted_line_price is not None:
+            price_with_discounts = (
+                line_price_with_discounts
+                if line_price_with_discounts is not None
+                else undiscounted_line_price
+            )
+            return OrderTaxedPricesData(
+                undiscounted_price=undiscounted_line_price,
+                price_with_discounts=price_with_discounts,
+            )
 
         return base_value
```

Comment:
wouldn't it return TaxedMoney instead of OrderTaxedPricesData as type in line 497 suggests?